lib/tasks/data_migration.rake

```diff
@@ -1,78 +1,61 @@
 namespace :data_migration do
   desc "Switch immediate subscribers of the specified list slugs to daily digest"
   task switch_to_daily_digest: :environment do |_t, args|
-    lists = SubscriberList.where(slug: args.extras)
-    raise "One or more lists were not found" if lists.count != args.extras.count
+    list_ids = SubscriberList.where(slug: args.extras).pluck(:id)
+    raise "One or more lists were not found" if list_ids.count != args.extras.count
 
-    subscriptions = Subscription.active.immediately.where(subscriber_list: lists)
+    subscriptions = Subscription.active.immediately.where(subscriber_list_id: list_ids)
     raise "No subscriptions to change" if subscriptions.none?
 
-    subscribers = Subscriber.where(id: subscriptions.pluck(:subscriber_id)).index_by(&:id)
-    subscriptions_by_subscriber = subscriptions.group_by(&:subscriber_id).transform_keys { |k| subscribers[k] }
-
-    subscriptions_by_subscriber.each do |subscriber, immediate_subscriptions|
-      email_id = nil
+    subscribers = Subscriber.where(id: subscriptions.pluck(:subscriber_id))
+
+    subscribers.find_in_batches(batch_size: 1000).with_index do |subscriber_batch, index|
+      puts "Processing batch #{index}"
+
+      subscriptions_by_subscriber = subscriptions
+        .where(subscriber: subscriber_batch)
+        .includes(:subscriber, :subscriber_list)
+        .group_by(&:subscriber)
+
+      subscriptions_by_subscriber.each do |subscriber, immediate_subscriptions|
+        email_id = nil
+        now = Time.zone.now
+
+        subscriber.with_lock do
+          new_subscriptions = immediate_subscriptions.map do |subscription|
+            {
+              subscriber_id: subscription.subscriber_id,
+              subscriber_list_id: subscription.subscriber_list_id,
+              frequency: :daily,
+              source: :bulk_immediate_to_digest,
+              created_at: now,
+              updated_at: now,
+            }
+          end
+
+          Subscription.where(id: immediate_subscriptions).update_all(
```

Comment:
This looks like there's a type mismatch where you're missing a pluck?

Reply:
ActiveRecord was converting them automatically. It's not a query anymore, so no `pluck`. Still good to be explicit, though 👍.

Reply:
Oh cool TIL that you can do that.